<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# Проект «Анализ вакансий из HeadHunter»
   

In [3]:
import pandas as pd
import psycopg2

In [4]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python
# dbname = public.project_sql

параметры подключения

In [5]:
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [6]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

## 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в базе (вакансии находятся в таблице `vacancies`).

In [6]:
# текст запроса
query_3_1 = f'''
    SELECT
        Count(id)
    FROM vacancies
'''

In [9]:
# результат запроса
df_3_1 = pd.read_sql_query(query_3_1, connection)
df_3_1

<ipython-input-9-7f364a65e0f6>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица `employers`).

In [11]:
# текст запроса
query_3_2 = f'''
SELECT
Count(id)
FROM employers
'''

In [12]:
# результат запроса
df_3_2 = pd.read_sql_query(query_3_2, connection)
df_3_2

<ipython-input-12-e0e2d23c59c3>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитайте с помощью запроса количество регионов (таблица `areas`).

In [13]:
# текст запроса
query_3_3 = f'''
SELECT
Count(id)
FROM areas
'''

In [14]:
# результат запроса
df_3_3 = pd.read_sql_query(query_3_3, connection)
df_3_3

<ipython-input-14-11c9948ba1f3>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитайте с помощью запроса количество сфер деятельности в базе (таблица `industries`).

In [15]:
# текст запроса
query_3_4 = f'''
SELECT
Count(id)
FROM industries
'''

In [16]:
# результат запроса
df_3_4 = pd.read_sql_query(query_3_4, connection)
df_3_4

<ipython-input-16-aae953b7cc87>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

In [ ]:
# выводы по предварительному анализу данных

По результатам предварительного анализа данных База данным project_sql состоит из 4 основных таблиц- справочники вакансий, регионов, работодателей и сфер дефтельности работодателей. Так же есть дополнительная таблица, которая необходима для осуществеления связи manytomany между таблицами работодателей и сфер их деятельности. Данная база данных нужна для составления списка вакансий работадателей.

 Предварительный анализ данных датасета позволил сделать следующие выводы:
1.	Количество вакансий в базе данных (вакансии находятся в таблице vacancies) – 49197.
2.	Количество работодателей (таблица employers) – 23501.
3.	Количество регионов (таблица areas) – 1362.
4.	количество сфер деятельности работадателей в базе (таблица industries) – 294.


## 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (`cnt`) вакансий в каждом регионе (`area`).
Отсортируйте по количеству вакансий в порядке убывания.

In [17]:
# текст запроса
query_4_1 = f'''
SELECT
a.name as area,
Count(v.id) as cnt
FROM vacancies v
     join areas a on v.area_id = a.id
Group by a.id
order by Count(v.id) desc
Limit 5
'''

In [18]:
# результат запроса
df_4_1 = pd.read_sql_query(query_4_1, connection)
df_4_1

<ipython-input-18-e0d1ca36e390>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_1 = pd.read_sql_query(query_4_1, connection)


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [20]:
# текст запроса
query_4_2 = f'''
SELECT
    COUNT(*) AS cnt
FROM
    vacancies v
WHERE
    v.salary_from IS NOT NULL OR
    v.salary_to IS NOT NULL;
'''

In [21]:
# результат запроса
df_4_2 = pd.read_sql_query(query_4_2, connection)
df_4_2

<ipython-input-21-a68fde1d0c96>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_2 = pd.read_sql_query(query_4_2, connection)


cnt
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до **целого числа**.

In [22]:
# текст запроса
query_4_3 = f'''
SELECT
    ROUND(AVG(salary_from)) AS avg_salary_from,
    ROUND(AVG(salary_to)) AS avg_salary_to
FROM
    vacancies v
'''

In [23]:
# результат запроса
df_4_3 = pd.read_sql_query(query_4_3, connection)
df_4_3

<ipython-input-23-5722848d6806>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_3 = pd.read_sql_query(query_4_3, connection)


avg_salary_from  avg_salary_to
0          71065.0       110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (`schedule`) и типа трудоустройства (`employment`), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [24]:
# текст запроса
query_4_4 = f'''
SELECT
    schedule,
    employment,
    COUNT(*) AS vacancy_count
FROM
    vacancies
GROUP BY
    schedule,
    employment
ORDER BY
    vacancy_count DESC
'''

In [25]:
# результат запроса
df_4_4 = pd.read_sql_query(query_4_4, connection)
df_4_4

<ipython-input-25-f3d798c9d673>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_4 = pd.read_sql_query(query_4_4, connection)


schedule           employment  vacancy_count
0        Полный день     Полная занятость          35367
1   Удаленная работа     Полная занятость           7802
2      Гибкий график     Полная занятость           1593
3   Удаленная работа  Частичная занятость           1312
4     Сменный график     Полная занятость            940
5        Полный день           Стажировка            569
6     Вахтовый метод     Полная занятость            367
7        Полный день  Частичная занятость            347
8      Гибкий график  Частичная занятость            312
9        Полный день     Проектная работа            141
10  Удаленная работа     Проектная работа            133
11     Гибкий график           Стажировка            116
12    Сменный график  Частичная занятость            101
13  Удаленная работа           Стажировка             64
14     Гибкий график     Проектная работа             18
15    Сменный график           Стажировка             12
16    Вахтовый метод     Проектная работа              2
17    Сменный график     Проектная работа              1

5. Напишите запрос, выводящий значения поля «Требуемый опыт работы» (`experience`) в порядке возрастания количества вакансий, в которых указан данный вариант опыта.

In [26]:
# текст запроса
query_4_5 = f'''
SELECT
    experience,
    count(id) as vacancy_count
FROM
    vacancies
GROUP BY

    experience
ORDER BY
    vacancy_count
'''

In [27]:
# результат запроса
df_4_5 = pd.read_sql_query(query_4_5, connection)
df_4_5

<ipython-input-27-a407e816ef92>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_5 = pd.read_sql_query(query_4_5, connection)


experience  vacancy_count
0         Более 6 лет           1337
1           Нет опыта           7197
2       От 3 до 6 лет          14511
3  От 1 года до 3 лет          26152

***

In [ ]:
# выводы по детальному анализу вакансий

**Детальный анализ датасета позволил сделать следующие выводы**:

1.	Наибольшее количество вакансий сосредоточено в городах-миллионниках: Москве, Санкт-Петербурге, Минске, Новосибирске, Алматы.
2.	У более, чем 50 %  вакансий не заполнено ни одно из полей по зарплате.
3.	Средняя зарплатная вилка по всей выборки составила от 71065 руб. до 110537 руб.
4.	Наиболее популярное  сочетание типа рабочего графика  и типа трудоустройства, используемого в вакансиях – это Полный день — Полная занятость. На втором месте сочетание Удалённая работа — Полная занятость.
5.	Наибольшее количество вакансии в выборке с  требуемым опытом работы От 1 года до 3 лет.


## 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [28]:
# текст запроса
query_5_1 = f'''
SELECT
   e.name,
    count(v.id) as vacancy_count
FROM
    vacancies v
    join employers e on v.employer_id = e.id
group by e.id
ORDER BY
    vacancy_count desc
limit 5
'''

In [29]:
# результат запроса
df_5_1 = pd.read_sql_query(query_5_1, connection)
df_5_1

<ipython-input-29-e71d17c95279>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_1 = pd.read_sql_query(query_5_1, connection)


name  vacancy_count
0         Яндекс           1933
1     Ростелеком            491
2       Тинькофф            444
3           СБЕР            428
4  Газпром нефть            331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [27]:
# текст запроса
query_5_2 = f'''
SELECT
   a.name,
   count(distinct e.id) as employer_count,
   count(distinct v.id) as vacancy_count
FROM
    areas a
    left join employers e on  a.id = e.area
    left join vacancies v on a.id = v.area_id

group by a.id
order by 3, 2 desc
'''

In [28]:
# результат запроса
df_5_2 = pd.read_sql_query(query_5_2, connection)
df_5_2

<ipython-input-28-14e6adc3ead1>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_2 = pd.read_sql_query(query_5_2, connection)


name  employer_count  vacancy_count
0                 Россия             410              0
1              Казахстан             207              0
2     Московская область              75              0
3     Краснодарский край              19              0
4     Ростовская область              18              0
...                  ...             ...            ...
1357              Алматы             721           1892
1358         Новосибирск             573           2006
1359               Минск            1115           2112
1360     Санкт-Петербург            2217           2851
1361              Москва            5864           5333

[1362 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [29]:
# текст запроса
query_5_3 = f'''
SELECT
    e.name,
    COUNT(DISTINCT v.area_id) AS area_count
FROM
    employers e
    JOIN vacancies v ON e.id = v.employer_id
GROUP BY
    e.id
ORDER BY
    area_count DESC
'''

In [30]:
# результат запроса
df_5_3 = pd.read_sql_query(query_5_3, connection)
df_5_3

<ipython-input-30-e83115aae06f>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_3 = pd.read_sql_query(query_5_3, connection)


name  area_count
0                                      Яндекс         181
1                                  Ростелеком         152
2                                  Спецремонт         116
3                      Поляков Денис Иванович          88
4                                   ООО ЕФИН           71
...                                       ...         ...
14901                            НПП Авиатрон           1
14902              Центр дистанционных торгов           1
14903  Городские Телекоммуникационные Системы           1
14904                       Введенский, Отель           1
14905              Группа Компаний Автоимпорт           1

[14906 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности.

In [5]:
# текст запроса
query_5_4 = f'''
SELECT
  Count(distinct e.id)
FROM
    employers e
    left join employers_industries t on e.id = t.employer_id
where t.industry_id is null
'''

In [6]:
# результат запроса
df_5_4 = pd.read_sql_query(query_5_4, connection)
df_5_4

<ipython-input-6-59cd48778173>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_4 = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности.

In [7]:
# текст запроса
query_5_5 = f'''
SELECT
    e.name
FROM
    employers e
    JOIN employers_industries t ON e.id = t.employer_id
GROUP BY
    e.id, e.name
HAVING
    COUNT(DISTINCT t.industry_id) = 4
ORDER BY
    e.name
OFFSET 2 LIMIT 1
'''

In [8]:
# результат запроса
df_5_5 = pd.read_sql_query(query_5_5, connection)
df_5_5

<ipython-input-8-55a59fdce066>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_5 = pd.read_sql_query(query_5_5, connection)


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана «Разработка программного обеспечения».


In [9]:
# текст запроса
query_5_6 = f'''
SELECT
    count(t.employer_id)
FROM
     employers_industries t
Where
    t.industry_id = (
    SELECT
    i.id
    From industries i
    Where i.name = 'Разработка программного обеспечения')
'''

In [10]:
# результат запроса
df_5_6 = pd.read_sql_query(query_5_6, connection)
df_5_6

<ipython-input-10-b8cbf15ac659>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_6 = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список [городов-миллионников](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8), в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку "Total" с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

    Если возникнут трудности с этим заданием, посмотрите материалы модуля 6.4 «Как получать данные из веб-источников и API».

In [20]:
# код для получения списка городов-милионников
import pandas as pd
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8#%D0%A1%D0%BF%D0%B8%D1%81%D0%BE%D0%BA_%D0%B3%D0%BE%D1%80%D0%BE%D0%B4%D0%BE%D0%B2-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D0%BE%D0%B2'
data = pd.read_html(url)
city_df = data[0]
city_list = list(city_df['Город'])
print(city_list)

# Преобразуем список городов в строку для SQL
city_list_str = ', '.join(f"'{city}'" for city in city_list)
print(city_list_str)

['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Самара', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград']
'Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Самара', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград'


In [32]:
# текст запроса
query_5_7 = f'''
SELECT
    COALESCE(a.name, 'Total') AS area_name,
    COUNT(v.id) AS vacancy_count
FROM
    vacancies AS v
JOIN
    employers AS e ON e.id = v.employer_id
JOIN
    areas AS a ON v.area_id = a.id
WHERE
    e.name = 'Яндекс' AND
    a.name IN ({city_list_str})
GROUP BY
    ROLLUP(a.name)
ORDER BY
    vacancy_count;
'''

In [33]:
# результат запроса
df_5_7 = pd.read_sql_query(query_5_7, connection)
df_5_7

<ipython-input-33-0bbb2c17c37c>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_7 = pd.read_sql_query(query_5_7, connection)


area_name  vacancy_count
0              Омск             21
1         Челябинск             22
2        Красноярск             23
3         Волгоград             24
4             Пермь             25
5            Казань             25
6    Ростов-на-Дону             25
7            Самара             26
8               Уфа             26
9         Краснодар             30
10          Воронеж             32
11      Новосибирск             35
12  Нижний Новгород             36
13     Екатеринбург             39
14  Санкт-Петербург             42
15           Москва             54
16            Total            485

***

In [ ]:
# выводы по анализу работодателей

Анализ работодателей позволил сделать следующие выводы:
1.	В топ-5 по количеству вакансий вошли следующие работодатели:
- Яндекс;
- Ростелеком;
- Тинькофф
- СБЕР;
- Газпром нефть.
2.	Наибольшее количество вакансий от работодателей сосредоточены в крупных городах: Москва, Санкт-Петербург, Минск и пр.
3.	В том работодателей по  наибольшему числу регионов, в которых опубликовали вакансии, вошли Яндекс и Ростелеком.
4.	В компании Яндекса представлены 485 вакансий в городах- миллионников, причем наибольшее число вакансий опубликовано для Москвы и Санкт-Петербурга.

## 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

    Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова `'data'` или `'данн'`.

    *Обратите внимание, что названия вакансий могут быть написаны в любом регистре.*


In [5]:
# текст запроса
query_6_1 = f'''
SELECT
    COUNT(*) AS vacancy_count
FROM
    vacancies v
WHERE
    lower(v.name) LIKE '%data%' OR
    lower(v.name) LIKE '%данн%'
'''

In [6]:
# результат запроса
df_6_1 = pd.read_sql_query(query_6_1, connection)
df_6_1

<ipython-input-6-e8ca8a3733ce>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_1 = pd.read_sql_query(query_6_1, connection)


vacancy_count
0           1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
    * 'data scientist'
    * 'data science'
    * 'исследователь данных'
    * 'ML' (здесь не нужно брать вакансии по HTML)
    * 'machine learning'
    * 'машинн%обучен%'

    **В следующих заданиях мы продолжим работать с вакансиями по этому условию.**

    Считаем вакансиями для специалистов уровня Junior следующие:
    + в названии есть слово “junior” **или**
    + требуемый опыт — «Нет опыта» **или**
    + тип трудоустройства — «Стажировка».


In [15]:
# текст запроса
query_6_2 = f'''
SELECT
    COUNT(*) AS suitable_vacancy_count
FROM
    vacancies v
WHERE
    (lower(v.name) LIKE '%data scientist%' OR
     lower(v.name) LIKE '%data science%' OR
     lower(v.name) LIKE '%исследователь данных%' OR
     (lower(v.name) LIKE '%ml%' AND lower(v.name) NOT LIKE '%html%') OR
     lower(v.name) LIKE '%machine learning%' OR
     lower(v.name) LIKE '%машинн%обучен%')
AND
    (lower(v.name) LIKE '%junior%' OR
     lower(v.experience) = 'нет опыта' OR
     lower(v.employment) = 'стажировка')
  '''

In [16]:
# результат запроса
df_6_2 = pd.read_sql_query(query_6_2, connection)
df_6_2

<ipython-input-16-b7a554361891>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_2 = pd.read_sql_query(query_6_2, connection)


suitable_vacancy_count
0                      51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или Postgres?

    *Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [1]:
# текст запроса
query_6_3 = f'''
SELECT
    COUNT(*) AS vacancy_count
FROM
    vacancies v
WHERE
    (lower(v.name) LIKE '%data scientist%' OR
     lower(v.name) LIKE '%data science%' OR
     lower(v.name) LIKE '%исследователь данных%' OR
     (lower(v.name) LIKE '%ml%' AND lower(v.name) NOT LIKE '%html%') OR
     lower(v.name) LIKE '%machine learning%' OR
     lower(v.name) LIKE '%машинн%обучен%')

AND
    (lower(v.key_skills) LIKE '%sql%' OR
     lower(v.key_skills) LIKE '%postgres%')
'''

In [6]:
# результат запроса
df_6_3 = pd.read_sql_query(query_6_3, connection)
df_6_3

<ipython-input-6-8a4199a81e3b>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_3 = pd.read_sql_query(query_6_3, connection)


vacancy_count
0            229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS. Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

    *Это можно сделать помощью запроса, аналогичного предыдущему.*

In [7]:
# текст запроса
query_6_4 = f'''
SELECT
    COUNT(v.id) AS vacancy_count
FROM
    vacancies v
WHERE
    (lower(v.name) LIKE '%data scientist%' OR
     lower(v.name) LIKE '%data science%' OR
     lower(v.name) LIKE '%исследователь данных%' OR
     (lower(v.name) LIKE '%ml%' AND lower(v.name) NOT LIKE '%html%') OR
     lower(v.name) LIKE '%machine learning%' OR
     lower(v.name) LIKE '%машинн%обучен%')

AND
    lower(v.key_skills) LIKE '%python%'
'''

In [8]:
# результат запроса
df_6_4 = pd.read_sql_query(query_6_4, connection)
df_6_4

<ipython-input-8-95116ecfece2>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_4 = pd.read_sql_query(query_6_4, connection)


vacancy_count
0            357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до **двух знаков после точки-разделителя**.

In [17]:
# текст запроса
query_6_5 = f'''
SELECT
    ROUND(AVG(LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), '')) + 1), 2) AS avg_key_skills,
    COUNT(*) AS total_vacancies
FROM public.vacancies
WHERE (LOWER(name) LIKE '%data scientist%'
   OR LOWER(name) LIKE '%data science%'
   OR LOWER(name) LIKE '%исследователь данных%'
   OR (LOWER(name) LIKE '% ml %' AND LOWER(name) NOT LIKE '%html%')
   OR LOWER(name) LIKE '%machine learning%'
   OR LOWER(name) LIKE '%машинн%обучен%')
   AND key_skills IS NOT NULL
   AND key_skills != '';
'''

In [18]:
# результат запроса

df_6_5 = pd.read_sql_query(query_6_5, connection)
df_6_5

<ipython-input-18-52923328eb0b>:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_5 = pd.read_sql_query(query_6_5, connection)


avg_key_skills  total_vacancies
0            6.41              352

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в среднем указывают для каждого типа требуемого опыта (уникальное значение из поля `experience`).

    При решении задачи примите во внимание следующее:
    1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
    2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
    3. Если в расчётах участвует `null`, в результате он тоже даст `null` (посмотрите, что возвращает запрос `select 1 + null`). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит `null` на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

    Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до **целого числа**.

In [25]:
# текст запроса
query_6_6 = f'''
SELECT
    experience,
    Round(AVG(COALESCE((salary_from + salary_to) / 2.0, salary_from, salary_to)), 0) AS average_salary
FROM
    vacancies
WHERE
    (LOWER(name) LIKE '%data scientist%'
   OR LOWER(name) LIKE '%data science%'
   OR LOWER(name) LIKE '%исследователь данных%'
   OR (name like '%ML%' and not name like '%HTML%')
   OR LOWER(name) LIKE '%machine learning%'
   OR LOWER(name) LIKE '%машинн%обучен%')
   AND key_skills IS NOT NULL
   AND key_skills != ''
GROUP BY
    experience
ORDER BY
    experience;
'''

In [26]:
# результат запроса
df_6_6 = pd.read_sql_query(query_6_6, connection)
df_6_6

<ipython-input-26-12f33d1ad98b>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_6 = pd.read_sql_query(query_6_6, connection)


experience  average_salary
0         Более 6 лет             NaN
1           Нет опыта         74643.0
2  От 1 года до 3 лет        146704.0
3       От 3 до 6 лет        243115.0

***

In [ ]:
# выводы по предметному анализу

Предметный анализ вакансий на позицию Data Scientist позволил сделать следующие выводы:
1.	Всего  вакансий, имеющих отношение к данным, в выборке составляет 1771 вакансий.
2.	Для начинающего специалиста в этой области представлена 51 вакансия.
3.	Причем  для DS, в которых в качестве ключевого навыка указан SQL или Postgres, представлено вакансий 229 вакансий.
4.	А вакансий для DS, в которых в качестве ключевого навыка указан Python, 357.
5.	Причем в среднем специалист DS должен  владеть 6 ключевыми навыками.
6.	Анализ средней зарплаты в зависимости от опыта работу показал, что наибольшую среднюю зарплату предлагают специалистам с опытом работы от 3 до 6 лет.


## Общий вывод по проекту

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

На основе предоставленных данных и анализа вакансий, можно сделать следующие общие выводы о состоянии рынка труда в сфере IT, в частности по вакансиям Data Scientist, а также по всем вакансиям в базе данных.

### 1. Анализ общего состояния вакансий

#### a) Общее количество вакансий
Общее количество вакансий в базе данных составляет 49197, что указывает на активный рынок труда в сфере IT. Избыточное предложение вакансий может быть связано с ростом IT-компаний, увеличением спроса на автоматизацию процессов и цифровизацию бизнеса.

#### b) Географическая распределенность
Наиболее активные регионы вакансий — это крупные города, такие как Москва, Санкт-Петербург, Минск, Новосибирск и Алматы. Это говорит о том, что такие города являются центрами концентрации IT-специалистов и работодателей. Высокий уровень работодателей в этих городах может быть объяснен большим количеством бизнесов, стартапов и IT-компаний, которые ищут квалифицированных специалистов.

#### c) Занятость и различные графики
Наиболее распространенные сочетания вакансий — это "Полный день" и "Полная занятость". Это свидетельствует о том, что большинство соискателей предпочитают стабильные формы занятости, которые часто предлагают более выгодные условия и прочные контракты. Однако растущий интерес к удаленной работе (например, "Удаленная работа — Полная занятость") также показывает изменение в подходах к работе, возможно, в связи с последствиями пандемии.

### 2. Работодатели и их активность
- Самыми крупными работодателями в базе данных, такими как Яндекс и Ростелеком, очевидно, требуется много вакансий для удовлетворения растущего спроса. Однако 8419 работодателей имеют неопределенные сферы деятельности, что может указывать на необходимость более четкой классификации и понимания собственных нужд в работе.
- Наличие 3553 работодателей в сфере "Разработка программного обеспечения" также подтверждает растущий интерес к данной области и необходимость квалифицированных рабочих ресурсов.

### 3. Анализ рынка труда
- На вакансии для специалистов Data Scientist составляет всего 1771, что говорит о более узкой специализации в данной области. Данный факт может указывать на высокую конкуренцию за такие вакансии, так как не так много открытых позиций относительно общего количества вакансий.
- Интересно отметить, что вакансии Data Scientist с указанием навыка SQL или Python составляют 229 и 357 соответственно, что говорит о том, что эти навыки являются самыми востребованными на рынке.

### 4. Заработная плата и опыт
Анализ заработной платы показывает значительную разницу в средних значениях в зависимости от опыта работы. Наибольшая зарплата для специалистов с опытом от 3 до 6 лет может быть следствием того, что именно в этом диапазоне работники становятся высококвалифицированными и способны выполнять более сложные задачи.

### 5. Возможные направления для будущих исследований
- **Анализ влияния образования на уровень зарплаты** — можно проанализировать, насколько образование (степень, сертификаты) связано с уровнем заработной платы, что поможет понять, ценятся ли дипломы в данной области.
- **Сравнительный анализ вакансий по регионам** — выявление наиболее перспективных регионов для трудоустройства в IT, а также декомпозиция на основе уровня жизни, затрат на жилье и пр.
- **Тенденции в культурных и социальных изменениях** — как изменения в методах работы (удаленная работа, гибкий график) влияют на поведение соискателей и их предпочтения.

### Заключение
В целом, данные показывают активный рынок вакансий в крупных городах, рост интереса к IT-сфере и потребность в специалистах с определенными навыками. Открытые вакансии Data Scientist свидетельствуют о высоком уровне конкуренции, а также о необходимости постоянного обновления знаний в быстро развивающейся отрасли. Направленные исследования в этой области могут помочь усовершенствовать процесс подбора кадров и дать более четкое представление о текущих и будущих трендах в области IT.

 Чтобы провести более детализированное исследование и анализ вакансий для соискателей в области Data Science можно предложить следующие:

- **Сравнительный анализ зарплат по регионам**: можно исследовать, как уровень заработной платы для позиций Data Scientist различается в разных регионах. Это поможет соискателям выбирать наиболее выгодные для себя места для работы.



In [13]:
vacancies = '''(
  (v.name like '%ML%' and not v.name like '%HTML%') or
  (lower(v.name) like '%data scientist%') or
  (lower(v.name) like '%data science%') or
  (lower(v.name) like '%исследователь данных%') or
  (lower(v.name) like '%machine learning%') or
  (lower(v.name) like '%машинн%обучен%')
)'''

In [14]:
query_7_1 = f'''
SELECT
    v.area_id,
    a.name,
    Round(AVG(COALESCE((v.salary_from + v.salary_to) / 2.0, v.salary_from, v.salary_to)), 0) AS average_salary
FROM
    vacancies v
    join areas a on v.area_id = a.id

WHERE
    {vacancies}
   and (v.salary_from is not null or v.salary_to is not null)
GROUP BY
    v.area_id, a.name
ORDER BY
    average_salary desc;
'''

In [15]:
df_7_1 = pd.read_sql_query(query_7_1, connection)
df_7_1

<ipython-input-15-19aad9e474fe>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_7_1 = pd.read_sql_query(query_7_1, connection)


area_id             name  average_salary
0       236             Кипр        300000.0
1        13          Армения        268863.0
2      5046       Черногория        233794.0
3       146           Сербия        233794.0
4        94           Турция        233794.0
5         1           Москва        207630.0
6        17         Белгород        200000.0
7         2  Санкт-Петербург        173335.0
8         4      Новосибирск        166333.0
9        77           Рязань        136667.0
10     2090          Королев        130000.0
11        3     Екатеринбург        124000.0
12      160           Алматы         94587.0
13       99              Уфа         90000.0
14       22      Владивосток         85000.0
15       72            Пермь         75000.0
16       11          Барнаул         60000.0
17       88           Казань         50000.0
18       76   Ростов-на-Дону         50000.0

Таким образом, наиболее выгодные по зарплате места работы для соискателя позиции  Data Scientist - это регионы Кипр, Армения, Черногория, Сербия, Москва.

- **Исследование количества вакансий для позиции Data Scientist по крупным городам**

Интересным будет исследование количества вакансий для позиции Data Scientist в городах-миллиониках самого крупного работодателя - Яндекса.

In [23]:
query_7_2 = f'''
SELECT
    COALESCE(a.name, 'Total') AS area_name,
    COUNT(v.id) AS vacancy_count
FROM
    vacancies AS v
JOIN
    employers AS e ON e.id = v.employer_id
JOIN
    areas AS a ON v.area_id = a.id
WHERE
    {vacancies} and
    e.name = 'Яндекс' AND
    a.name IN ({city_list_str})
GROUP BY
    ROLLUP(a.name)
ORDER BY
    vacancy_count;
'''

In [24]:
df_7_2 = pd.read_sql_query(query_7_2, connection)
df_7_2

<ipython-input-24-84fc3f6e799c>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_7_2 = pd.read_sql_query(query_7_2, connection)


area_name  vacancy_count
0     Екатеринбург              1
1      Новосибирск              1
2           Казань              1
3  Санкт-Петербург              2
4           Москва              4
5            Total              9

Как видно в Яндексе очень мало вакансий для Data Scientist.

Такая ситуация требует более тщательного и детального анализа вакансий по топ-10 работодателям.

---



In [31]:
query_7_2_1 = f'''
SELECT
    e.name,
    COUNT(*) AS vacancy_count
FROM
    employers e
    JOIN vacancies v ON e.id = v.employer_id

Where
   (v.name like '%ML%' and not v.name like '%HTML%') or
  (lower(v.name) like '%data scientist%') or
  (lower(v.name) like '%data science%') or
  (lower(v.name) like '%исследователь данных%') or
  (lower(v.name) like '%machine learning%') or
  (lower(v.name) like '%машинн%обучен%')

GROUP BY
   e.id
ORDER BY
    vacancy_count DESC
'''

In [32]:
df_7_2_1 = pd.read_sql_query(query_7_2_1, connection)
df_7_2_1

<ipython-input-32-8ef57178c35c>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_7_2_1 = pd.read_sql_query(query_7_2_1, connection)


name  vacancy_count
0                                                 СБЕР             37
1                                      Bell Integrator             25
2                                       Банк ВТБ (ПАО)             18
3                                                   VK             15
4                                Positive Technologies             11
..                                                 ...            ...
220  Федеральное автономное учреждение Государствен...              1
221  Представительство ООО «Квалитет» (Российская Ф...              1
222                                         DIGINETICA              1
223                                               Дром              1
224                                              Sifox              1

[225 rows x 2 columns]

Таким образом, наиболее заинтересованы в специалистах Data Scientist - это СБЕР, Bell Integrator, Банк ВТБ (ПАО), VK.

Также можно посмотреть сколько всего вакансий предлагается по городам-миллионикам.

In [25]:
query_7_3 = f'''
SELECT
    COALESCE(a.name, 'Total') AS area_name,
    COUNT(v.id) AS vacancy_count
FROM
    vacancies AS v
JOIN
    employers AS e ON e.id = v.employer_id
JOIN
    areas AS a ON v.area_id = a.id
WHERE
    {vacancies} and
    a.name IN ({city_list_str})
GROUP BY
    ROLLUP(a.name)
ORDER BY
    vacancy_count;
'''

In [26]:
df_7_3 = pd.read_sql_query(query_7_3, connection)
df_7_3

<ipython-input-26-7895489308b8>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_7_3 = pd.read_sql_query(query_7_3, connection)


area_name  vacancy_count
0        Красноярск              1
1         Волгоград              1
2         Челябинск              1
3           Воронеж              2
4               Уфа              2
5              Омск              3
6            Самара              5
7             Пермь              5
8         Краснодар              6
9    Ростов-на-Дону              6
10     Екатеринбург              7
11           Казань             16
12  Нижний Новгород             20
13      Новосибирск             23
14  Санкт-Петербург             64
15           Москва            217
16            Total            379

Таким образом, на российском IT-рынке труда не так много вакансий для Data Scientist, всего 379, и 57% из них сосредоточены в Москве.

- **Удаленная работа по регионам**: провести анализ числа вакансий с удаленной работой в различных регионах.

In [29]:
query_7_4 = f'''
SELECT
    v.schedule,
    v.employment,
    COUNT(*) AS vacancy_count
FROM
    vacancies v
Where
   {vacancies}
   and v.schedule = 'Удаленная работа'
GROUP BY
    v.schedule,
    v.employment
ORDER BY
    vacancy_count DESC
'''

In [30]:
df_7_4 = pd.read_sql_query(query_7_4, connection)
df_7_4

<ipython-input-30-5892e389fb30>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_7_4 = pd.read_sql_query(query_7_4, connection)


schedule           employment  vacancy_count
0  Удаленная работа     Полная занятость            110
1  Удаленная работа  Частичная занятость              3
2  Удаленная работа     Проектная работа              2

Из запроса видно, что вакансии с удаленной работой для  Data Scientist составляют примерно 24 % от всего числа вакансий для   Data Scientist.

### **Итоги**:

Таким образом, соискатель должен для успешного поиска работы должен обладать в среднем 6 ключевыми навыками, из которых главные - это знание SQL или Postgres, Python.
Понимать, более востребованные специалисты с опытом работы  от 3 до 6 лет. В зависимости от опыта работы рассчитывать на среднюю зарплату для джуна 75000 руб, для мидла- 145000 руб., для синьора – 240000 руб.

По локации наиболее привлекательными по зарплате являются регионы Кипр, Армения, Черногория, Сербия, Москва.

 В основном требуются специалисты на полный рабочий день и полную занятость. Но можно найти и удаленную работу.

Работодатели, которые заинтересованы в специалистах  Data Scientist - это СБЕР, Bell Integrator, Банк ВТБ (ПАО), VK.
